docs/architecture/adr-031-typed-events.md

```diff
@@ -0,0 +1,421 @@
+# ADR 030: Typed Events
+
+## Changelog
+
+- 28-Sept-2020: Initial Draft
+
+## Authors
+
+- Anil Kumar (@anilcse)
+- Jack Zampolin (@jackzampolin)
+- Adam Bozanich (@boz) 
+
+## Status
+
+Proposed
+
+## Context
+
+Currently in the SDK, events are not properly organized (event data is defined in the handlers for each message) and are prone to inconsistencies. This makes customizing events difficult. In addition, they are difficult to consume. This proposal focuses on updating the events to use **typed events** in each module such that emiting and subscribing to events will be much more egrnomic. This will make use of the SDK to build event driven processes much easier and enable rapid development of things like relayers and automated transaction bots for SDK based chains. These types of bots enable will enable easy building of new features at exchanges, wallets, explorers, and defi protocols. IBC especially will benefit from this proposal.
```

Comment:
I'm sorry, I'm sure this has great motivation and after reading this, I like the idea of where it's going, but the context here doesn't do much in terms of understanding why this proposal is being made.

1. WDYM not organized well? Yes, events are just simple type/attribute tuples, how are they not organized well and how does this cause issues for you?
2. How are they prone to inconsistencies? Please provide explicit examples.
3. How is customizing difficult? Events are not meant to be customized, but rather added on-top of. Please be explicit here too.
4. Events are already "typed", I think you mean they're wrapped with context.

Reply:
Your work on events to organize and streamline them got us to the place where we could imagine building a typed pubsub system on top of them to make for easy consumption. The events are currently just `map[string]string` pairs appended to the events. This is mainly due to the limitations in the tendermint event system. This ADR primarily aims to offer golang SDK developers types to organize their events. This allows for much easier parsing and avoids exposing developers to raw strings. See [this `relayer` code](https://github.com/ovrclk/relayer/blob/master/relayer/naive-strategy.go#L166) which is error prone and hard to debug.

I'm sorry if the wording here feels pointed or otherwise unfair. Happy to change that however you would like. What I think is the core value add here is an extremely easy to use system for consuming events via the tendermint websocket subscription that will make building event driven applications like the [`relayer`](https://github.com/ovrclk/relayer), [`provider-daemon`](https://github.com/ovrclk/akash/tree/master/provider), [`deploy`](https://github.com/ovrclk/deploy), `peggy-relayer`, and other future oracle or peg applications extremely easy and extensible. This turns all of these applications into a specific set of event handlers that are easy to debug, maintain and extend.

Reply:
I've completely rewritten this section to be much more actual `Context`. Thank you very much for the feedback @alexanderbez